### DISCUSSION TITLE
The case for Zig

### DISCUSSION THREAD

**Discussion by v1gnesh** [zopencommunity/meta#350] [Ideas]
I know, we're still talking about Rust.
However, it's worth also keeping tabs on Zig.

From [File for Divorce from LLVM](https://github.com/ziglang/zig/issues/16270)
> We can attract direct contributions from Intel, ARM, RISC-V chip manufacturers, etc., who have a vested interest in making our machine code better on their CPUs.

- [x] Name starts in Z. Case closed.

Can use different allocators for different functions, based on the requirement - https://ziglearn.org/chapter-2/#allocators
Can use different build modes for different functions, allowing different levels of optimizations.
[Data-oriented design](https://vimeo.com/649009599). Imagine this and all the various data-processing use cases on zOS.
No more ~~tears~~ pre-processor macros.

**Comment by IgorTodorovskiIBM** [comment]
> I know, we're still talking about Rust. However, it's worth also keeping tabs on Zig.
> 
> From [File for Divorce from LLVM](https://github.com/ziglang/zig/issues/16270)
> 
> > We can attract direct contributions from Intel, ARM, RISC-V chip manufacturers, etc., who have a vested interest in making our machine code better on their CPUs.
> 
> * [x]  Name starts in Z. Case closed.
> 
> Can use different allocators for different functions, based on the requirement - https://ziglearn.org/chapter-2/#allocators Can use different build modes for different functions, allowing different levels of optimizations. [Data-oriented design](https://vimeo.com/649009599). Imagine this and all the various data-processing use cases on zOS. No more ~tears~ pre-processor macros.

I'm impressed by the Bun.js performance numbers (bun.js is written in zig for those who are unaware).  This article provide a pretty good summary of zig: https://www.infoworld.com/article/3689648/meet-the-zig-programming-language.html. The one negative I've heard is that there's no official package manager and that the language is changing rapidly. It looks like they plan to release version 1.0 in 2025.

**Reply by MikeFultonDev** [reply]
Does look interesting. Would be neat to have zig working in z/OS from the start

**Reply by v1gnesh** [reply]
> The one negative I've heard is that there's no official package manager and that the language is changing rapidly. It looks like they plan to release version 1.0 in 2025.

Yeah, it's being worked on right now. https://github.com/ziglang/zig/pull/14265
Staying in sync with Zig's development means we benefit from the first principles thinking that's being applied to its package manager, build system, self-contained tooling, etc.
Bun and [Tigerbeetle](https://tigerbeetle.com/) are astonishing.
You can feel the focus, clarity, and brutal simplicity in Zig's project planning, communication and execution.

I'm not even a programmer but I'm very curious about what the Z wizards think about Zig's custom memory allocators, and what it unlocks for Z. Maybe revealing my cluelessness as I say this, but I'm wondering if this can help move more and more stuff from below the line/bar to 64-bit.
